feat(hogql): Add retention appearance to persons query runner

webjunkie opened the pull request:
## Problem

As follow-up from #18831 we need to still port the persons/actors modal queries to HogQL.

## Changes

- add to_persons_query to retention query runner
- enhance persons query runner
- enhanced pagination


https://github.com/PostHog/posthog/assets/59713/b0ccc1bd-5f1c-4705-bcc2-d11fb1d7d192



## How did you test this code?

- added the tests that where skipped in #18831 
- tested in frontend and compared

webjunkie committed:
Add retention query runner for actor appearances

webjunkie committed:
Merge branch 'master' into feature/retention-actors-hogql

webjunkie committed:
Make groups work

webjunkie committed:
Make work with PersonsQuery

webjunkie committed:
Make pagination nicer

webjunkie committed:
Handle load more

webjunkie committed:
Add explore functionality

webjunkie committed:
Fix pagination results trim

webjunkie committed:
Bring back tests

webjunkie committed:
Refactor paginator

webjunkie committed:
Fix frontend types

webjunkie committed:
Fix types

webjunkie committed:
Revert explore button to put in follow-up PR

webjunkie commented on posthog/hogql_queries/insights/retention_appearance_query_runner.py:
```diff
@@ -0,0 +1,140 @@
+        limit_context: Optional[LimitContext] = None,
+    ):
+        super().__init__(query, team=team, timings=timings, modifiers=modifiers, limit_context=limit_context)
+        self.paginator = HogQlHasMorePaginator(limit=self.query_limit(), offset=self.query.offset or 0)
```
I would make other query runners also use this paginator in a separate PR.

webjunkie commented on posthog/hogql_queries/insights/retention_appearance_query_runner.py:
```diff
@@ -0,0 +1,140 @@
+        return appearances_query
+
+    def get_actors_from_result(self, raw_result) -> Union[List[SerializedGroup], List[SerializedPerson]]:
+        # TODO: Overhaul by leveraging common functionality in PersonsQueryRunner
```
As discussed, I would want to have common functionality for that somewhere. Would be a follow-up PR.

webjunkie commented on posthog/hogql_queries/insights/retention_appearance_query_runner.py:
```diff
@@ -0,0 +1,140 @@
+            hogql=response.hogql,
+            **self.paginator.response_kwargs(),
+        )
+        #   "missing_persons": len(actor_appearances) - len(actors),
```
@mariusandra I guess this number is still important? Would it be alright to have a field on `PersonsQueryResponse` with `extended_data` or something?

mariusandra replied:
I honestly don't know the context behind this. It was probably something to do with us getting tired of "why is my data off by a few" messages by users. I don't know how used it is, but if we could have it somehow, it would be great.

webjunkie replied:
I added it as field to persons response, since when persons are returned we potentially could always miss some I guess.

webjunkie reviewed (commented):
Some notes.

webjunkie committed:
Merge branch 'master' into feature/retention-actors-hogql

mariusandra commented on posthog/hogql_queries/insights/paginators.py:
```diff
@@ -0,0 +1,62 @@
+from posthog.schema import HogQLQueryResponse
+
+
+class HogQlHasMorePaginator:
```
Offtopic: I changed the pagination logic recently for the persons runner. Now instead of +offset, it doubles the results every time. Each time we make one query to ClickHouse, it's costly, and we re-read the same data, even if we discard the bottom 99% of it. Doubling the results every time actually spares our servers.

webjunkie replied:
This logic is in the frontend, right? It's supported by my paginator since it keeps accepting offsets and limits.

mariusandra commented on posthog/hogql_queries/insights/retention_appearance_query_runner.py:
```diff
@@ -0,0 +1,140 @@
+                breakdown_values_filter=[self.query.selectedInterval]
+            ),
+        }
+        with self.timings.measure("retention_query"):
+            retention_query = parse_select(
+                """
+                    SELECT
+                        actor_id,
+                        groupArray(actor_activity.intervals_from_base) AS appearances
+
+                    FROM {actor_query} AS actor_activity
+
+                    GROUP BY actor_id
+
+                    -- make sure we have stable ordering/pagination
+                    -- NOTE: relies on ids being monotonic
+                    ORDER BY
+                        length(appearances) DESC,
+                        actor_id
+                """,
+                placeholders,
+                timings=self.timings,
+            )
+        return retention_query
```
This could likely go into the `to_actors_query` method inside `RetentionQueryRunner`

mariusandra reviewed (changes requested):
Great work! However thinking about the queries a bit, and considering what I need in order to make the TODO list item "session recording info in persons modal" a success, I think we could streamline a few things. 

For the session recording task (look at any trends person modal --> it lists also the recordings in the time range), I need to get the field `select person_id, properties.$session_id from events`, and pass that through `PersonsQuery`.

To make the analysis complete, for the `LifecycleQuery`'s `InsightPersonsQuery`, I would also like to pass on the breakdown options (`day`, `status`) to the end users.

Thus the `InsightPersonsQuery` when running will have to become something like:

```sql
select person_id, properties.$session_id, day, status from (...)
```

This feels very similar to what the `RetentionAppearanceQuery` is doing:

```sql
select person_id, week_array from (...)
```

The `PersonsQuery` can then JOIN this with data with the actual `persons` table, group by person_id, and return the week or session data to the users.

Thus the steps I'd take are:

1. Rename `InsightPersonsQuery` to `InsightActorsQuery`, and `to_persons_query` to `to_actors_query` everywhere.
2. Merge `RetentionAppearanceQuery` into `InsightActorsQuery`
3. Make the query return all the relevant fields for each insight type.
4. Update `PersonsQuery` to swap the `id in ({insight_persons_query})` with an inner join.

This means a query that ultimately looks like:

```ts
{
  kind: "PersonsQuery",
  select: ["person", "source.week_1", "source.week_2"]
  source: {
    kind: "InsightActorsQuery",
    interval: 3,
    source: { kind: "RetentionQuery", ... }
  }
}
```

I'm not sure at the end of the day how easy/practical this is, so push back if you have other thoughts!

webjunkie commented on frontend/src/queries/schema.ts:
```diff
@@ -757,14 +768,29 @@ export type InsightFilter =
+export interface RetentionAppearanceQuery extends PaginatableQuery {
+    kind: NodeKind.RetentionAppearanceQuery
+    source: RetentionQuery
+    /** @asType integer */
+    selectedInterval: number
```
I think I would rather have the selected interval be part of the retention query, meaning you can get the persons off of it only when this is provided. Otherwise we could end up adding too much unrelated things to persons query that are more the concern of the source query.

webjunkie committed:
Merge branch 'master' into feature/retention-actors-hogql

webjunkie committed:
Try to use persons query runner more

webjunkie committed:
Make query work returning appearances as well

webjunkie committed:
Swap to new in frontend

webjunkie committed:
Adjust tests to use PersonsQuery

webjunkie committed:
Adjust appearance results in frontend and adjust tests

webjunkie committed:
Add back missing persons/actors

webjunkie committed:
Remove retention appearance runner again

Since we use persons query runner now for that

webjunkie committed:
Use new paginator in persons query runner

webjunkie committed:
Update actors enrichment

webjunkie committed:
Bring explore button already

webjunkie committed:
Refactor frontend

webjunkie committed:
Fix one test problem

webjunkie committed:
Move explore button

webjunkie committed:
Fix type

webjunkie commented:
I've moved the functionality from the retention appearance runner into the retention runner that now works better with the persons query runner and utilizes the `to_persons_query` really well. I think this is a nice approach.

I also adjusted the persons query runner to use join instead of using `IN`. There are a couple of intricacies with that, but I think we can solve it. One remaining issue for example is how it should work for groups, if this is at all desirable for the persons query runner. Or do we then wrap it with a different one maybe? It's needed so far to support the persons modal in the retention query when you have groups instead of people selected. I've made a shortcut for now so that we don't needlessly join with the person table in that case, but obviously all filtering and such on the persons query is bypassed in that case 😅 

I've also brought to this PR the `Explore` button, which works really nice.

The renaming to `InsightActorsQuery` I would do in a clean follow-up PR.

webjunkie committed:
Refactor

webjunkie committed:
Merge branch 'master' into feature/retention-actors-hogql

mariusandra committed:
fix some tests

mariusandra committed:
fix few tests, add clearly breaking test

mariusandra reviewed (commented):
Overall this is really great work! 

I added a few changes to make some tests pass. Will dig into what's the case with the person_id error in a bit. Submitting my review now to already give the comments I had. (Breaking for lunch... 😋)

I think the only change I'd still do is to move the `selected_interval` into the `InsightPersonsQuery`. I believe we'll have a better time overall this way. The reason is that as a standalone query, this parameter has no impact, which will be confusing to users:

![image](https://github.com/PostHog/posthog/assets/53387/8f753217-71b5-4742-b0f5-5fcc156d201c)

(Still shows lines/results for all intervals)

One other reason why I wanted to put all of these options to InsightPersonsQuery, leaving the original insight untouched, is so that we could easily go back to the original insight if needed:

![image](https://github.com/PostHog/posthog/assets/53387/9232444b-b12e-4246-bc29-4d5405888fc8)

Alternatively, I can also see people wanting to embed just the retention line graph, and have just one selected interval visible instead of all of them. This way we could keep the filter like it's now, but then we should make it filter results properly and give people the option to configure it in the interface somehow. This feels like more work to me than moving the parameter to the `InsightPersonsQuery`...

mariusandra commented:
One more quick comment.

One of the failing tests: https://github.com/PostHog/posthog/actions/runs/7207631068/job/19634865044?pr=19211

... shows that the PersonsQuery got a bunch of new returned fields.

![image](https://github.com/PostHog/posthog/assets/53387/8d03df02-167d-4c4d-98cb-4551223ccd80)

I wonder if they all need to come back for all queries. The `uuid` and `id` columns seem redundant. The `value_at_data_point` field will be empty for most `PersonQuery`-s (e.g. just the `/persons` page). Not sure if those keys should/could be omitted by default, and only presented if anything to show there?

webjunkie commented on frontend/src/scenes/retention/RetentionModal.tsx:
```diff
@@ -90,8 +102,8 @@ export function RetentionModal(): JSX.Element | null {
                                 </tr>
-                                {people.result &&
-                                    people.result.map((personAppearances: RetentionTableAppearanceType) => (
+                                {people.results &&
+                                    people.results.map((personAppearances: RetentionTableAppearanceType) => (
```
Ah good catch, I wanted to just rename to `results` but might have not updated it somewhere.

webjunkie committed:
Move modal buttons

webjunkie committed:
Fix and rename

webjunkie committed:
Join the groups table nicely

webjunkie committed:
Hide search and property filter for groups

webjunkie committed:
Move selected_interval

webjunkie committed:
Fix tests

webjunkie committed:
Fix types

webjunkie committed:
Add strategies for persons query runner

webjunkie committed:
Refactor

webjunkie committed:
Improve persons search labels

webjunkie committed:
Push down figuring group_type_index

webjunkie committed:
Simplify by using length(appearances) when querying

webjunkie committed:
Handle order by in strategy as well

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

webjunkie committed:
Merge branch 'master' into feature/retention-actors-hogql

webjunkie committed:
Fix types

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

mariusandra committed:
Merge branch 'master' into feature/retention-actors-hogql

mariusandra commented on frontend/src/queries/nodes/PersonsNode/PersonsSearch.tsx:
```diff
@@ -4,38 +4,57 @@ import { Tooltip } from 'lib/lemon-ui/Tooltip'
     setQuery?: (query: PersonsNode | PersonsQuery) => void
 }
 
+type actorType = 'person' | 'group'
```
```suggestion
type ActorType = 'person' | 'group'
```

The convention so far has been to capitalize types.

webjunkie replied:
Typo, will fix.

mariusandra commented on frontend/src/queries/nodes/PersonsNode/PersonsSearch.tsx:
```diff
@@ -4,38 +4,57 @@ import { Tooltip } from 'lib/lemon-ui/Tooltip'
 
+type actorType = 'person' | 'group'
+
+const labels: Record<actorType, any> = {
```
Let's avoid `any`-s

```suggestion
interface LabelType {
    label: string
    description: string
}

const labels: Record<ActorType, LabelType> = {
```

mariusandra commented on frontend/src/queries/nodes/PersonsNode/PersonsSearch.tsx:
```diff
@@ -4,38 +4,57 @@ import { Tooltip } from 'lib/lemon-ui/Tooltip'
-                    </>
-                }
-            >
+            <Tooltip title={<>{labels[target].description}</>}>
```
nit: can be simplified a bit:
```suggestion
            <Tooltip title={labels[target].description}>
```

webjunkie replied:
I think I just moved that around... maybe the title needs to be an element? I'll just leave it as it is.

mariusandra commented on frontend/src/queries/nodes/PersonsNode/PersonsSearch.tsx:
```diff
@@ -4,38 +4,57 @@ import { Tooltip } from 'lib/lemon-ui/Tooltip'
+    return (
+        isPersonsQuery(query) &&
+        isInsightPersonsQuery(query.source) &&
+        isRetentionQuery(query.source.source) &&
+        query.source.source.aggregation_group_type_index !== undefined
```
Not blocking for this PR, but this means we'll need to know to edit this specific file when adding new query types that also aggregate on groups.

```
        isRetentionQuery(query.source.source) &&
        query.source.source.aggregation_group_type_index !== undefined
```

To clean this a bit, this could probably get its own `isQueryWithGroupAggregation(query.source.source)` util.

webjunkie replied:
That's a bit tricky since a lot of type checking gets in the way (source is only present on specific queries and so on and so on). I moved the whole check to a util function.

mariusandra commented on frontend/src/queries/schema.ts:
```diff
@@ -765,6 +776,11 @@ export interface InsightPersonsQuery {
+     * An interval selected out of available intervals in source query
+     * @asType integer
+     */
+    selected_interval?: number
```
I wonder if in the context of this query, `selected_` is redundant? Both "day" and "status" above are also technically "selected".

Might be worth simplifying to just `interval` while we still can?

webjunkie replied:
Uh yeah I like that.

mariusandra reviewed (commented):
Had some moments at the end of the day to review. Didn't get to the python part sadly, but dumping the comments I had.

Something's still up with the `person_id` alias in the tests as well, even after the fix in master :/.

webjunkie committed:
Merge branch 'master' into feature/retention-actors-hogql

webjunkie committed:
Rename selected_interval to interval in types

webjunkie commented:
> Something's still up with the `person_id` alias in the tests as well, even after the fix in master :/.

As far as I can dig in, I think it's down to the join targeting the `person_id` column as `Field`, but deep down this is an expression for something on the events table with PoE2 (`PERSON_ON_EVENTS_V2_OVERRIDE=True`). Not sure how to address that 😩

mariusandra committed:
Merge branch 'feature/retention-actors-hogql' of github.com:PostHog/posthog into feature/retention-actors-hogql

mariusandra commented on frontend/src/scenes/retention/retentionModalLogic.ts:
```diff
@@ -43,6 +46,36 @@ export const retentionModalLogic = kea<retentionModalLogicType>([
+                    full: true,
+                }
+                if (
+                    isInsightPersonsQuery(personsQuery.source) &&
+                    isRetentionQuery(personsQuery.source.source) &&
+                    personsQuery.source.source.aggregation_group_type_index !== undefined
```
This can probably be swapped out for `isQueryForGroup()`, however why are you actually hiding the property filter here? Users should be able to filter the results further...

webjunkie replied:
As far as I saw it, the filter is geared towards persons, and if we have groups it's not applicable, hence this was my idea to remove it.

mariusandra replied:
Right, makes sense.

mariusandra commented on frontend/src/scenes/retention/queries.ts:
```diff
@@ -0,0 +1,51 @@
+    const response = await query(actorsQuery)
+    const results: RetentionTableAppearanceType[] = response.results.map((row) => ({
+        person: row[0],
+        appearances: appearances_1s_0s(row[1], retentionQuery.retentionFilter.total_intervals || 11, selectedInterval),
```
Could this be done at the SQL level? 🤔 
I think it'd be nicer and easier for the users, if they get a list/array/sequence of 0-s and 1-s  back directly from the query. That makes further analysis easier (`appearances[1] == true`)

webjunkie replied:
I somehow found it easier to understand that I get a list of intervals: I ask for week, total_intervals = 10, and selected interval 2 and get back [2,3,4]. A list of 1s and 0s or even booleans would IMO be harder to manage for someone wanting to use the data somewhere else.

mariusandra commented on posthog/hogql_queries/insights/paginators.py:
```diff
@@ -0,0 +1,62 @@
+from posthog.schema import HogQLQueryResponse
+
+
+class HogQLHasMorePaginator:
```
Should this get some explicit tests? 🤷

mariusandra reviewed (approved):
This is a lot of great work! I still have some ideas for future improvement (mainly returning an array of booleans from SQL directly), but nothing blocking us merging this now.

Except those PoE HogQL tests. Will dig in now, but leaving the ✅ regardless. 👍

mariusandra committed:
fix alias
